File: packages/llm-test-harness/src/metrics/similarity.ts
import type { MetricFn, MetricScore } from '../types.js'

function levenshtein(a: string, b: string): number {
  const m = a.length
  const n = b.length
  const dp: number[][] = []

  for (let i = 0; i <= m; i++) {
    dp[i] = []
    for (let j = 0; j <= n; j++) {
      if (i === 0) {
        dp[i]![j] = j
      } else if (j === 0) {
        dp[i]![j] = i
      } else if (a[i - 1] === b[j - 1]) {
        dp[i]![j] = dp[i - 1]![j - 1]!
      } else {
        dp[i]![j] = 1 + Math.min(dp[i - 1]![j]!, dp[i]![j - 1]!, dp[i - 1]![j - 1]!)
      }
    }
  }

  return dp[m]![n]!
}

/**
 * Normalized Levenshtein similarity: 1 - (editDistance / maxLength).
 * Returns 1 for identical strings, 0 for maximally different strings.
 */
export function normalizedSimilarity(a: string, b: string): number {
  if (a === b) return 1
  const maxLen = Math.max(a.length, b.length)
  if (maxLen === 0) return 1
  return 1 - levenshtein(a, b) / maxLen
}

export class Similarity implements MetricFn {
  readonly name = 'Similarity'

  constructor(
    private reference: string,
    private threshold = 0.8,
  ) {}

  evaluate(text: string): MetricScore {
    const score = normalizedSimilarity(text, this.reference)
    const pass = score >= this.threshold
    return {
      name: this.name,
      pass,
      score,
      reason: pass
        ? undefined
        : `Similarity ${score.toFixed(3)} is below threshold ${this.threshold}`,
    }
  }
}

File: packages/llm-test-harness-python/llm_test_harness/metrics/similarity.py
from __future__ import annotations

from ..types import MetricFn, MetricScore


def _edit_distance(a: str, b: str) -> int:
    m, n = len(a), len(b)
    prev = list(range(n + 1))
    for i in range(1, m + 1):
        curr = [i] + [0] * n
        for j in range(1, n + 1):
            if a[i - 1] == b[j - 1]:
                curr[j] = prev[j - 1]
            else:
                curr[j] = 1 + min(prev[j], curr[j - 1], prev[j - 1])
        prev = curr
    return prev[n]


def normalized_similarity(a: str, b: str) -> float:
    if not a and not b:
        return 1.0
    max_len = max(len(a), len(b))
    if max_len == 0:
        return 1.0
    return 1.0 - _edit_distance(a, b) / max_len


class Similarity(MetricFn):
    def __init__(self, reference: str, *, threshold: float = 0.8) -> None:
        self._reference = reference
        self._threshold = threshold

    @property
    def name(self) -> str:
        return "Similarity"

    def evaluate(self, text: str) -> MetricScore:
        score = normalized_similarity(text, self._reference)
        passed = score >= self._threshold
        reason = None if passed else f"Similarity {score:.3f} below threshold {self._threshold}"
        return MetricScore(name=self.name, passed=passed, score=score, reason=reason)
